lib/OpenQA/WebAPI/Controller/API/V1/Job.pm

```diff
@@ -663,8 +663,25 @@ sub done {
     $validation->optional('result');
     $validation->optional('reason');
     $validation->optional('newbuild')->num(1);
+    $validation->optional('worker_id')->num(0, undef);
     return $self->reply->validation_error({format => 'json'}) if $validation->has_error;
 
+    # check whether the specified worker matches the actually assigned one; refuse the update if not
+    if (my $specified_worker_id = $validation->param('worker_id')) {
+        my $assigned_worker_id = $job->assigned_worker_id;
+        if (defined $assigned_worker_id) {
+            return $self->render(
+                status => 400,
+                json => {error => "Refusing to set result because job is now assigned to worker $assigned_worker_id)."}
+            ) if $assigned_worker_id != $specified_worker_id;
+        }
+        else {
+            return $self->render(
+                status => 400,
+                json   => {error => 'Refusing to set result because job has been re-scheduled.'});
+        }
```

Comment:
how about

```suggestion
        my $msg = 'Refusing to set result because job'
        $msg .= (defined $assigned_worker_id && $assigned_worker_id != $specified_worker_id) ?
            " is now assigned to worker $assigned_worker_id)." : ' has been re-scheduled.'
        return $self->render(status => 400, json => {error => $msg});
```

Reply:
The code you've been suggested is actually wrong. I've now created a version which is a little bit longer but still saves the double render call and is still correct.

Reply:
It turned out not that nice though. Why not try it with the msg concatenation?

Reply:
There's a case when the early return must not be executed and hence no message should be concatenated at all. I'm wondering what the goal of further "beautifying" this code is. It already looks kind of ok to me. (My original version was actually easier to read and the only problem was some repetition.)